engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -7,23 +7,49 @@
 import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
 import org.enso.interpreter.runtime.data.ManagedResource;
 
 /** Allows the context to attach garbage collection hooks on the removal of certain objects. */
 public final class ResourceManager {
-  private final EnsoContext context;
-  private volatile boolean isClosed = false;
-  private volatile Thread workerThread;
-  private final ProcessItems worker = new ProcessItems();
+  /** Amount of milliseconds to wait for another resource when none is pending. */
+  private static final long KEEP_ALIVE = 1000;
+
+  /** Queue with resources eligible for finalization */
   private final ReferenceQueue<ManagedResource> referenceQueue = new ReferenceQueue<>();
-  private final ConcurrentMap<PhantomReference<ManagedResource>, Item> items =
-      new ConcurrentHashMap<>();
+
+  /**
+   * All the items that were issued, but haven't yet arrived at {@link #referenceQueue} for
+   * finalization.
+   *
+   * <p>@GuardedBy("this")
+   */
+  private final List<Item> pendingItems = new ArrayList<>();
+
+  private final EnsoContext context;
+
+  /**
+   * Indicates this manager is closed. If assigned to {@code processor} field, this manager is
+   * considered closed.
+   */
+  private final ProcessItems CLOSED = new ProcessItems(null);
+
+  /**
+   * The current processor (with own worker {@link Thread}) that is supposed to await {@link
+   * #removeNextQueuedItem(ProcessItems)} and process them by properly invoking their finalizers
+   * using a safepoint action.
+   *
+   * <p>If this field is set to {@link #CLOSED}, then this resource manager is considered to be
+   * closed.
+   *
+   * <p>@GuardedBy("this")
+   */
+  private ProcessItems processor;
```

Comment:
This is the biggest _"moving part"_ in the `ResourceManager`. The `processor` field is either:
- `CLOSED` - no new `Item` instances can be registered
- `null` - no `pendingItems` and (soon) no active thread processing them
- otherwise it points to active instance of `ProcessItems` with a thread waiting for `Item` from `referenceQueue`